test_bandwidth.py

```diff
@@ -0,0 +1,204 @@
+        dist.all_to_all(output_tensor_list, input_tensor_list)
+
+    def ranks_factor(self):
+        return (self.world_size - 1) / self.world_size 
+
+def avg_time(op, tensor:torch.tensor, iters):
+    torch.cuda.synchronize()
+    start = time.time()
+    for _ in range(iters):
+        if isinstance(op, Broadcast):
+            if dist.get_rank == 0:
+                src = dist.get_rank()
+                op(tensor, src)
+        else:
+            op(tensor)
+    torch.cuda.synchronize()
+    end = time.time()
+    return (end - start) / iters
+
+def test_bandwidth(op, tensor_list:List[int], iters: int = 10, warmup_iters: int = 5, dtype = torch.float, collective_op: str = 'allreduce'):
+    #warm up
+    mintensor = torch.rand(tensor_list[0], dtype=dtype, device='cuda')
+    maxtensor = torch.rand(tensor_list[-1], dtype=dtype, device='cuda')
+    avg_time(op, mintensor, warmup_iters)
+    avg_time(op, maxtensor, warmup_iters)
+    pretty_table = PrettyTable(['bytes(B)', 'tensorElements', 'dtype', 'collective op', 'time(ms)',
+                        'algbw(GB/s)', 'busbw(GB/s)'], float_format='.3')
+    busbw_total = 0
+    for tensor_elements in tensor_list:
+        tensor = torch.rand(tensor_elements, dtype=dtype, device='cuda')
+        time = avg_time(op, tensor, iters)
+        if isinstance(op, Broadcast):
+            time_per_node = time
+        else:
+            #total time
+            time = torch.tensor([time],device = 'cuda')
+            dist.all_reduce(time)
+            time_per_node = time/dist.get_world_size()
+        if hasattr(time_per_node, 'item'):
+            time_per_node = time_per_node.item()
+        #bandwidth
+        algbw = (tensor_elements * (torch.finfo(dtype).bits // 8))/ time_per_node
+        busbw = algbw * op.ranks_factor()
+        busbw_total += busbw
+        pretty_table.add_row([(tensor_elements * (torch.finfo(dtype).bits // 8)), tensor_elements,
+                              dtype, collective_op, time_per_node*1000, algbw / 1024**3, busbw / 1024**3])
+    if dist.get_rank() == 0:
+        avg_busbw = busbw_total / len(tensor_list)
+        print(pretty_table)
+        print(f'Average busbw: {avg_busbw/1024**3:.3f} GB/s')    
+
+def get_op(ops: str, world_size: int = 1)-> CollectiveOps:
+    if ops == 'allreduce':
+        return AllReduce(world_size)
+    elif ops == 'allgather':
+        return AllGather(world_size)
+    elif ops == 'broadcast':
+        return Broadcast(world_size)
+    elif ops == 'alltoall':
+        return AllToAll(world_size)
+
+def str_to_int(bytes: str) -> int:
+    if bytes[-1] == 'B':
+        bytes = bytes[:-1]
+    elif bytes[-1] == 'K':
+        return int(bytes[:-1]) * 1024
+    elif bytes[-1] == 'M':
+        return int(bytes[:-1]) * 1024**2
+    elif bytes[-1] == 'G':
+        return int(bytes[:-1]) * 1024**3
+    return int(bytes)
+
+def get_tensor_list(minbytes:str, maxbytes:str, stepbytes:str, stepfactor:str, dtype = torch.float) -> List[int]:
```

Comment:
user snake case for readability, e.g. stepfactor -> step_factor.

Reply:
![image](https://github.com/wangbluo/BandWidth_Test/assets/32676639/552d744f-d828-40da-8679-a13362fd9860)
https://github.com/NVIDIA/nccl-tests
I saw the stepbytes and stepfactor parameters in the nccl tests's read.me, so I just want to keep it same. I'll fix it.